### ISSUE TITLE
Serf unable to bind to network subinterface

### ISSUE THREAD

**Issue by charleswhchan** [hashicorp/serf#309]
Trying to setup mesh network using [B.A.T.M.A.N](http://www.open-mesh.org/projects/batman-adv/wiki/Quick-start-guide) with Avahi; and run serf on top of the mesh network. 

```
$ ifconfig
bat0      Link encap:Ethernet  HWaddr fa:f3:e1:24:5b:40
          inet6 addr: fe80::f8f3:e1ff:fe24:5b40/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:47703 errors:0 dropped:0 overruns:0 frame:0
          TX packets:46338 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:3891442 (3.7 MiB)  TX bytes:3769271 (3.5 MiB)

bat0:avahi Link encap:Ethernet  HWaddr fa:f3:e1:24:5b:40
          inet addr:169.254.6.206  Bcast:169.254.255.255  Mask:255.255.0.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1

eth0      Link encap:Ethernet  HWaddr 6c:ec:eb:ad:70:4a
          inet addr:192.168.3.115  Bcast:192.168.3.255  Mask:255.255.255.0
          inet6 addr: fe80::6eec:ebff:fead:704a/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:3712330 errors:0 dropped:0 overruns:0 frame:0
          TX packets:139834 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:357287184 (340.7 MiB)  TX bytes:10875553 (10.3 MiB)
          Interrupt:40
```

By default, serf will try to bind to eth0/192.168.x.x address. So following the instructions under ["Configuration"](https://www.serfdom.io/docs/agent/options.html), I tried to use add the `-iface` option, but this results in an error:

```
$ ./serf agent -iface=bat0:avahi -discover=my-cluster
==> Invalid network interface: no such network interface
```

Not sure if this is a problem in serf or Go network library? Is this a valid use case for `-iface`?

**Comment by charleswhchan** [comment]
The current workaround is to get the IP of the `bat0:avahi` interface using:
   ```ifconfig bat0:avahi | grep 'inet addr:' | cut -d: -f2 | awk '{ print $1}'```

**Comment by slackpad** [comment]
Hmm - it's hard to tell what's going on here. It would be interesting to
write a little test program that calls `Interfaces()` and see what it says
on that system - http://golang.org/pkg/net/#Interfaces. That's the data
structure that ends up getting searched by Serf.

**Comment by charleswhchan** [comment]
@slackpad: I am not too familiar with Go programming. Think you can give me a "hello world" program to run? (That will save me a bit of time ... )

**Comment by slackpad** [comment]
Sure - this should do it:

```go
package main

import (
        "fmt"
        "net"
)

func main() {
        ifcs, err := net.Interfaces()
        if err != nil {
                panic(err)
        }

        for _, ifc := range ifcs {
                fmt.Println(ifc.Name)
        }
}
```

You can save that as interface.go and then run `go build interface.go`, which should dump out the executable in the same directory.

**Comment by charleswhchan** [comment]
Thanks! Ran the test program:

```
$ ./interface
lo
eth0
ra0
bat0
usb0
```

So it seems like it's not listed in the list returned by `net.Interfaces()`. Not sure how to proceed next? Report an issue in Go?

**Comment by charleswhchan** [comment]
By the way, I should mention I have also tried the primary `bat0` interface, but it gave me this other error:
```
$ ./serf agent -iface=bat0 -discover=my-cluster
==> Failed to find usable address for interface 'bat0'
```
Not sure what it means, as an inet6 address does exist.

**Comment by slackpad** [comment]
Hmm - it's not super clear where the problem is because Go is getting info from a system call and filtering it. I think I'd want to try to set up a better repro case and then try to see the output of the system call to see if it's a Go thing, or maybe even a BATMAN thing. Can you summarize how I could run a simple environment that does this (in a VM, for example). If it's a lot to set up then I'd probably give you a patch to try to collect the info.

That last issue seems like a bug, but could be related because that code is also looking over the interfaces.

**Comment by charleswhchan** [comment]
(My setup uses BeagleBone Black...)
At a high level, to replicate in a Debian/Ubuntu VM you will need:

1. Use `apt-get` install batctl and batman-adv kernel module:
http://www.open-mesh.org/projects/open-mesh/wiki/Download

2. Load batman-adv kernel module: `modprobe batman-adv`

3. Add eth0 (or other interface) to batman: `batctl if add eth0`
http://www.open-mesh.org/projects/batman-adv/wiki/Quick-start-guide

5. Start bat0 interface: `ifconfig bat0 up`

6. Check bat0 interface is setup by running `ifconfig`.

Let me know if you need further assistance with the setup; or I can run the patches to gather more facts.

**Comment by charleswhchan** [comment]
@slackpad: Were you able to get this going? Need any help?

**Comment by slackpad** [comment]
@charleswhchan: Sorry for the delay - I've been a little swamped with stuff. I'm hoping to take a look early next week. Your instructions look super clear so I think I should be able to give it a go.

**Comment by charleswhchan** [comment]
@slackpad: No problem! Thanks for the update and feel free to ping me if needed.

**Comment by charleswhchan** [comment]
@slackpad Any update on this? :relaxed:

**Comment by slackpad** [comment]
Sorry I got bogged down with things but I finally had a chance to try this! I ran a Ubuntu-based Vagrant box and then got the interfaces into the same state:

```
vagrant@test:~$ uname -a
Linux test 3.16.0-30-generic #40~14.04.1-Ubuntu SMP Thu Jan 15 17:43:14 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

vagrant@test:~$ ifconfig
bat0      Link encap:Ethernet  HWaddr 66:34:77:ee:30:48
          inet6 addr: fe80::6434:77ff:feee:3048/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:2 errors:0 dropped:0 overruns:0 frame:0
          TX packets:5 errors:0 dropped:8 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:84 (84.0 B)  TX bytes:210 (210.0 B)

bat0:avahi Link encap:Ethernet  HWaddr 66:34:77:ee:30:48
          inet addr:169.254.7.21  Bcast:169.254.255.255  Mask:255.255.0.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1

eth0      Link encap:Ethernet  HWaddr 00:0c:29:55:37:c9
          inet addr:172.16.30.202  Bcast:172.16.30.255  Mask:255.255.255.0
          inet6 addr: fe80::20c:29ff:fe55:37c9/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:60461 errors:0 dropped:0 overruns:0 frame:0
          TX packets:30242 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:81601229 (81.6 MB)  TX bytes:2033627 (2.0 MB)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

vagrant@test:~$ ./interface
lo
eth0
bat0
```

Looks like the same result that you got, and this was with the latest Go 1.5.1. I ran `strace` on the Go test program from the comment above as well as `ifconfig` to see what they were doing to enumerate the interfaces.

`ifconfig`
```
socket(PF_INET, SOCK_DGRAM, IPPROTO_IP) = 4
ioctl(4, SIOCGIFCONF, {120, {{"lo", {AF_INET, inet_addr("127.0.0.1")}}, {"eth0", {AF_INET, inet_addr("172.16.30.202")}}, {"bat0:avahi", {AF_INET, inet_addr("169.254.7.21")}}}}) = 0
```

`interface.go`
```
socket(PF_INET6, SOCK_STREAM, IPPROTO_TCP) = 3
setsockopt(3, SOL_IPV6, IPV6_V6ONLY, [1], 4) = 0
bind(3, {sa_family=AF_INET6, sin6_port=htons(0), inet_pton(AF_INET6, "::1", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, 28) = 0
socket(PF_INET6, SOCK_STREAM, IPPROTO_TCP) = 4
setsockopt(4, SOL_IPV6, IPV6_V6ONLY, [0], 4) = 0
bind(4, {sa_family=AF_INET6, sin6_port=htons(0), inet_pton(AF_INET6, "::ffff:127.0.0.1", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, 28) = 0
close(4)                                = 0
close(3)                                = 0
socket(PF_NETLINK, SOCK_RAW, 0)         = 3
bind(3, {sa_family=AF_NETLINK, pid=0, groups=00000000}, 12) = 0
sendto(3, "\21\0\0\0\22\0\1\3\1\0\0\0\0\0\0\0\0", 17, 0, {sa_family=AF_NETLINK, pid=0, groups=00000000}, 12) = 17
recvfrom(3, "l\4\0\0\20\0\2\0\1\0\0\0\203\n\0\0\0\0\4\3\1\0\0\0I\0\1\0\0\0\0\0"..., 4096, 0, {sa_family=AF_NETLINK, pid=0, groups=00000000}, [12]) = 3448
getsockname(3, {sa_family=AF_NETLINK, pid=2691, groups=00000000}, [12]) = 0
getsockname(3, {sa_family=AF_NETLINK, pid=2691, groups=00000000}, [12]) = 0
getsockname(3, {sa_family=AF_NETLINK, pid=2691, groups=00000000}, [12]) = 0
recvfrom(3, "\24\0\0\0\3\0\2\0\1\0\0\0\203\n\0\0\0\0\0\0", 4096, 0, {sa_family=AF_NETLINK, pid=0, groups=00000000}, [12]) = 20
getsockname(3, {sa_family=AF_NETLINK, pid=2691, groups=00000000}, [12]) = 0
```

It looks like they are using different mechanisms to obtain this information. I'm a little shaky on what's going on but looking at the Go source code it looks like it's querying some kind of routing information from the RIB. What's interesting is that running `ip route` which is supposed to show the RIB's contents shows the `bat0` interface but with the `bat0:avahi` IPV4 address:

```
vagrant@test:~$ ip route
default via 172.16.30.2 dev eth0
default dev bat0  scope link  metric 1003
169.254.0.0/16 dev bat0  proto kernel  scope link  src 169.254.7.21
172.16.30.0/24 dev eth0  proto kernel  scope link  src 172.16.30.202
```

So at this point it seems like Go just uses a different mechanism to enumerate the interfaces that works differently. I don't think it's a bug, but it might be worth asking on the mailing list. Sorry it took so long to reproduce this, and that I don't have a better answer on this one.

**Comment by charleswhchan** [comment]
Thanks for the analysis @slackpad, will query Go mailing list ...

**Comment by slackpad** [comment]
I'm going to close this out since it seems like an issue with Go, and not Serf itself. Please reopen if there's anything we need to do on the Serf side for this.
